internal/mesh/internal/controllers/sidecarproxy/fetcher/data_fetcher.go

```diff
@@ -0,0 +1,255 @@
+package fetcher
+
+import (
+	"context"
+
+	"github.com/hashicorp/consul/internal/catalog"
+	"github.com/hashicorp/consul/internal/mesh/internal/cache/sidecarproxycache"
+	ctrlStatus "github.com/hashicorp/consul/internal/mesh/internal/controllers/sidecarproxy/status"
+	"github.com/hashicorp/consul/internal/mesh/internal/types"
+	intermediateTypes "github.com/hashicorp/consul/internal/mesh/internal/types/intermediate"
+	"github.com/hashicorp/consul/internal/resource"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type Fetcher struct {
+	Client pbresource.ResourceServiceClient
+	Cache  *sidecarproxycache.Cache
+}
+
+func (f *Fetcher) FetchWorkload(ctx context.Context, id *pbresource.ID) (*intermediateTypes.Workload, error) {
+	rsp, err := f.Client.Read(ctx, &pbresource.ReadRequest{Id: id})
```

Comment:
non-blocking:

One of my other PRs landed `resource.GetDecodedResource` to do pretty much all of this method body.

Reply:
the only thing it wouldn't do for you is your cache interaction and assembling your response object.

Reply:
Yeah I think I'll address it in a follow-up PR .